sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteWithExpression.scala

```diff
@@ -34,7 +34,7 @@ import org.apache.spark.sql.catalyst.trees.TreePattern.{COMMON_EXPR_REF, WITH_EX
  */
 object RewriteWithExpression extends Rule[LogicalPlan] {
   override def apply(plan: LogicalPlan): LogicalPlan = {
-    plan.transformWithPruning(_.containsPattern(WITH_EXPRESSION)) {
+    plan.transformDownWithSubqueriesAndPruning(_.containsPattern(WITH_EXPRESSION)) {
```

Comment:
why do we need this change?

Reply:
I think all subqueries will be optimized recursively. There is no exception.

Reply:
This is needed when an Aggregate (count-bug-susceptible) has a WITH expression in it, because we skip optimizing it in OptimizeSubqueries. This only matters because RewriteWithExpression appears very early on in the optimization, and it is not triggered after we unnest the subquery. Any "later" rule would apply on the rewritten subquery, so it would not be a problem. For this "early" rules there are two options: i) repeat them after the subquery is fully unnested (i.e., the count bug has been handled, and the subquery is replaced with a join), or ii) make sure they run on subqueries. I chose option ii).

Reply:
To be more precise, this PR deals with the following subqueries:
S:
Aggregate (mayhavecountbug = true)
 |
<subquery body>

Before this PR, OptimizeSubqueries applied on S, with this PR OptimizeSubqueries applies on <subquery body>. 
After we unnest the subquery, the rules apply on the entire query (including the aggregate), so we don't miss any optimization opportunities by excluding Aggregate in OptimizeSubqueries.
However, a rule RewriteWithExpressions is done before OptimizeSubqueries, so we need to make sure subqueries are handled there (in case Aggregate node has a WITH expresion in it).

Reply:
Then this rule becomes O(n^2) complexity as every level of subqueries runs this rule for all subqueries under its level.

Thinking about it more, why do we handle the count bug in two places that are far away?

Reply:
Why is it O(n^2)? The ReplaceWithExpression would replace all WITH expressions in one pass when we first call it, would not it? (I know that technically it is also called as part of optimization loop in OptimizeSubqueries, but by that point all the WITH expressions are replaced).